packages/dds/tree/src/test/tree/visitDelta.spec.ts

```diff
@@ -990,6 +990,155 @@ describe("visitDelta", () => {
+			const index = makeDetachedFieldIndex("", testRevisionTagCodec);
+			const buildId = { minor: 42 };
+			const rootFieldDelta: DeltaFieldChanges = {
+				local: [
+					{
+						count: 1,
+						fields: new Map([[fooKey, { local: [{ count: 1, attach: buildId }] }]]),
+					},
+				],
+			};
+			const expected: VisitScript = [
+				["enterField", rootKey],
+				["enterNode", 0],
+				["enterField", fooKey],
+				["exitField", fooKey],
+				["exitNode", 0],
+				["exitField", rootKey],
+				["enterField", rootKey],
+				["enterNode", 0],
+				["enterField", fooKey],
+				["create", [content], field0],
+				["attach", field0, 1, 0],
+				["exitField", fooKey],
+				["exitNode", 0],
+				["exitField", rootKey],
+			];
+			const delta: DeltaRoot = {
+				refreshers: [{ id: buildId, trees: [content] }],
+				fields: new Map([[rootKey, rootFieldDelta]]),
+			};
+			testVisit(delta, expected, index);
+			assert.equal(index.entries().next().done, true);
+		});
+
+		it("for partial restores", () => {
+			const index = makeDetachedFieldIndex("", testRevisionTagCodec);
+			const node = { minor: 42 };
+			const rootFieldDelta: DeltaFieldChanges = {
+				local: [{ count: 1, attach: { minor: 43 } }],
+			};
+			const delta: DeltaRoot = {
+				refreshers: [{ id: node, trees: [content, content] }],
+				fields: new Map([[rootKey, rootFieldDelta]]),
+			};
+			const expected: VisitScript = [
+				["enterField", rootKey],
+				["exitField", rootKey],
+				["enterField", rootKey],
+				["create", [content], field0],
+				["attach", field0, 1, 0],
+				["exitField", rootKey],
+			];
+			testVisit(delta, expected, index);
+			assert.equal(index.entries().next().done, true);
+		});
+
+		it("for changes to repair data", () => {
+			const index = makeDetachedFieldIndex("", testRevisionTagCodec);
+			const refresherId = { minor: 42 };
+			const buildId = { minor: 43 };
+			const rootFieldDelta: DeltaFieldChanges = {
+				global: [
+					{
+						id: refresherId,
+						fields: new Map([[fooKey, { local: [{ count: 1, attach: buildId }] }]]),
+					},
+				],
+			};
+			const expected: VisitScript = [
+				["create", [content], field0],
+				["enterField", rootKey],
+				["create", [content], field1],
+				["exitField", rootKey],
+				["enterField", field1],
+				["enterNode", 0],
+				["enterField", fooKey],
+				["exitField", fooKey],
+				["exitNode", 0],
+				["exitField", field1],
+				["enterField", rootKey],
+				["exitField", rootKey],
+				["enterField", field1],
+				["enterNode", 0],
+				["enterField", fooKey],
+				["attach", field0, 1, 0],
+				["exitField", fooKey],
+				["exitNode", 0],
+				["exitField", field1],
+			];
+			const delta: DeltaRoot = {
+				refreshers: [{ id: refresherId, trees: [content] }],
+				build: [{ id: buildId, trees: [content] }],
+				fields: new Map([[rootKey, rootFieldDelta]]),
+			};
+			testVisit(delta, expected, index);
+		});
+	});
+
+	it("tolerates superfluous refreshers", () => {
```

Comment:
There are two things that might make a refresher superfluous:
A. The delta can be applied without having the refreshed tree in the forest
B. The refreshed tree already exists in the forest

This test covers A. It would be good to have a test for B as well.